designs/sam-config.md

```diff
@@ -0,0 +1,413 @@
+
+If this could be condensed into a series of workflows that look like
+
+* `sam build`
+* `sam package`
+* `sam deploy`
+
+That would be a huge user experience win.
+
+Tenets
+-------------------------------
+
+* Resolution of command line parameters should always favor explicit versus implicit. A native command line parameter specified directly on the command line should override a parameter specified in the configuration file.
+
+What will be changed?
+---------------------
+
+The suite of commands supported by SAM CLI would be aided by looking for a configuration file thats locally located under the `.aws-sam/` at the project root where template.yaml is located by default.
+
+`.aws-sam/samconfig.toml`
+
+
+This configuration would be used for specifiying the parameters that each of SAM CLI commands use and would be in TOML format.
+
+Running a SAM CLI command now automatically looks for `.aws-sam/samconfig.toml` file and if its finds it goes ahead with parameter passthroughs to the CLI.
+
+```
+sam build
+Default Config file location: .aws-sam/samconfig.toml
+..
+..
+..
+```
+
+Why samconfig under `.aws-sam`
+---------------------------------
+
+> The below two don't answer the above question. Instead, they give technical implementation answers to questions not asked like "How is the .aws-sam directory made?" and "Are any files in .aws-sam git ignored?". The ideas surfaced in the previous PR's discussion were lost. For example, in the PR threads it was for a short time in the root of the project but then put back into .aws-sam. I recommend clearly stating here an answer to the above question in terms of goals, not technical implementations.
+
+The `.aws-sam` directory within the project directory is created with normal 755 permissions as default without any special permisions. `sam build` only creates a build directory within `.aws-sam` as `.aws-sam/build`. This directory is erased and re-built on every build. but top level directory is left unaffected.
+
+The `.gitignore` specified in the init apps also only have `.aws-sam/build` ignored and not anything else.
+
+
+Config file versioning
+-----------------------
+
+The configuration file: `samconfig.toml` will come with a top level version key that specifies the grammar version of the configuration file. This version can then be used to determine if a given configuration file works with a version of SAM CLI. A given version of SAM CLI may support one or more configuration file grammar versions.
+
+> The above clarifies that the version is not the version of the file itself (like an ever increasing commit version), instead it is the version of the grammar/DTD/spec of the file.
+
+It also paves the forward when major changes need to be made to the configuration file and add a version bump to the config file version
+
+```
+version = 0.1
+```
+
+> What is behavior of SAM CLI for different scenarios, e.g.:
```

Comment:
1 & 2 are covered in the implementation, whereas 3 is not. Can document what the behaviors will be.